TheOceanCleanupAlgorithms/ADVECT
# Update Boundary Behavior for 3D

- [x] vector decompose/sort
- [x] boundary condition update
- [x] verify no particles enter bathymetry (plot or output analysis)

Before, ocean boundary behavior was handled by a "slippery coastlines" sort of idea, wherin particles can't exit the domain, and instead try to scoot along the boundary.  This was performed by removing one of the displacement vector's components, with preference for removing the smaller components first.

This idea can extend to 3D, but is more complex to treat.  The ideas is still to try and modify the displacement vector by removing components, with preference for the smaller components, until the displacement does not shuttle the particle out of domain.

Now, there are 3 steps:
1. decompose the displacement vector into three orthogonal component vectors.  Sort these by magnitude.
2. Try removing a singular component, ascending in magnitude.
3. Try removing two components, ascending in magnitude.
4. Give up, don't move particle at all.

Step 1 turns out to be kind of gnarly to code in opencl. It's sorting an array (of vectors), which is already tricky in C, by the values in a separate array (vector magnitude), which means there's gonna be void *s and memcopys and all kinds of stuff I'd rather avoid.  So instead, since this is only an array of length 3, I do the if-else tree which covers all cases.  Actually not so bad, and I would guess is the fastest method too.

This is complex logic, and it's hard to test just by looking at advector output.  So finally, I've written a unit test.  I didn't try to make the testing structure scalable, I can deal with that when I eventually write more tests.
